src/app.rs

```diff
@@ -114,100 +91,49 @@ impl App {
 
         let database = Arc::new(database);
 
-        // Poseidon tree depth is one more than the contract's tree depth
-        let tree_state = Arc::new(TimedRwLock::new(
-            Duration::from_secs(options.lock_timeout),
-            TreeState::new(
-                identity_manager.tree_depth() + 1,
-                identity_manager.initial_leaf_value(),
-            ),
-        ));
+        let tree_state = Self::initialize_tree(
+            &database,
+            identity_manager.tree_depth() + 1,
+            identity_manager.initial_leaf_value(),
+        )
+        .await?;
 
         let identity_committer = Arc::new(IdentityCommitter::new(
             database.clone(),
             identity_manager.clone(),
             tree_state.clone(),
         ));
-        let chain_subscriber = EthereumSubscriber::new(
-            options.starting_block,
-            database.clone(),
-            identity_manager.clone(),
-            tree_state.clone(),
-        );
```

Comment:
With the remover of this, it'd be great if you can remove the `fetch_events` method from the `IdentityManager` trait. That's been a pain and not having to deal with it would be wonderful.